File: interview-75/29/main.go
package main

func spiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return []int{}
	}
	row, column := len(matrix), len(matrix[0])
	visited := make([][]bool, row)
	for i := range visited {
		visited[i] = make([]bool, column)
	}

	total := row * column
	directions := [][]int{
		[]int{0, 1},
		[]int{1, 0},
		[]int{0, -1},
		[]int{-1, 0},
	}
	directIdx := 0
	r, c := 0, 0
	result := make([]int, total)

	for i := 0; i < total; i++ {
		result[i] = matrix[r][c]
		visited[r][c] = true
		nextRow, nextCol := r+directions[directIdx][0], c+directions[directIdx][1]
		if (nextCol < 0 || nextCol >= column || nextRow < 0 || nextRow >= row) || (visited[nextRow][nextCol]) {
			directIdx = (directIdx + 1) % 4
		}
		r += directions[directIdx][0]
		c += directions[directIdx][1]
	}

	return result
}

func main() {
	m := [][]int{}
	spiralOrder(m)
}

File: interview-75/29/main.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix or not matrix[0]:
            return []

        row, column = len(matrix), len(matrix[0])
        left, right, top, bottom = 0, column-1, 0, row-1
        result = []
        while left <= right and top <= bottom:
            for c in range(left, right+1):
                print(top, c)
                result.append(matrix[top][c])

            for r in range(top+1, bottom+1):
                print(r, right)
                result.append(matrix[r][right])

            if left < right and top < bottom:
                for c in range(right-1, left, -1):
                    print(bottom, c)
                    result.append(matrix[bottom][c])

                for r in range(bottom, top, -1):
                    print(r, left)
                    result.append(matrix[r][left])

            left, right, top, bottom = left+1, right-1, top+1, bottom-1

        return result


s = Solution()
m = [[3, 2]]
r = s.spiralOrder(m)
print(r)
